crates/ruff_db/src/files.rs

```diff
@@ -87,33 +87,70 @@ impl Files {
             .system_by_path
             .entry(absolute.clone())
             .or_insert_with(|| {
-                tracing::trace!("Adding file '{path}'");
-
                 let metadata = db.system().path_metadata(path);
-                let durability = self
-                    .root(db, path)
-                    .map_or(Durability::default(), |root| root.durability(db));
-
-                let builder = File::builder(FilePath::System(absolute))
-                    .durability(durability)
-                    .path_durability(Durability::HIGH);
-
-                let builder = match metadata {
-                    Ok(metadata) if metadata.file_type().is_file() => builder
-                        .permissions(metadata.permissions())
-                        .revision(metadata.revision()),
-                    Ok(metadata) if metadata.file_type().is_directory() => {
-                        builder.status(FileStatus::IsADirectory)
-                    }
-                    _ => builder
-                        .status(FileStatus::NotFound)
-                        .status_durability(Durability::MEDIUM.max(durability)),
-                };
 
-                builder.new(db)
+                self.create_system_entry(db, absolute, metadata)
             })
     }
 
+    /// Looks up or inserts a [`File`] at the given path using the provided metadata.
+    ///
+    ///
+    /// Prefer using [`system_path_to_file`] unless its `stats` call when inserting a new file is a performance
+    /// concern because the metadata is already available or when computing the metadata on different threads.
+    ///
+    /// The metadata is only used when inserting a new file. It doesn't
+    /// update the metadata if a [`File`] for `path` already exists.
+    ///
+    /// This function differs from [`system_path_to_file`] in that it doesn't add a dependency
+    /// on `File.status` and always returns a `File`, even
+    /// if `path` doesn't point to a file (e.g. if `path` is a directory).
```

Comment:
How come? If `metadata` lets you cheaply query the file type, can you still return a `Result` here?

I mention this because it seems like we are otherwise careful not to allow a `File` to be backed by a directory.